### PR TITLE
rw table: add value prev

### PR DESCRIPTION
**Pull request by lispc**
fixing https://github.com/privacy-scaling-explorations/zkevm-specs/issues/202

Only reversible rws have `value_prev`.
There is no need to constain MemoryRw and StackRw since the 'read
consistency' part of the constaints are enough for them to behave
correctly.

For these 6 Rws whose `value_prev` need to be
constrained:

1. `AccountStorage` and `Account`: they are related to storage
and they should be connected to MPT cricuit later to check the
`value_prev`.
2. `TxAccessListAccount` and
`TxAccessListAccountStorage`:  Default values of them should be `false`
indicating "not accessed yet".
3.  `AccountDestructed`: Since we probably
will not support this feature, it is skipped now.
4. `TxRefund`: Default values should be '0'. BTW it may be moved out of rw table in the future. See https://github.com/privacy-scaling-explorations/zkevm-circuits/issues/395
for more details.

### CODE REVIEW

File: src/zkevm_specs/state.py [hunk 1]
```diff
@@ -328,8 +333,11 @@ def check_tx_refund(row: Row, row_prev: Row):
     assert row.field_tag() == 0
     assert row.storage_key() == 0
 
+    # 7.1 When keys change, value_prev must be 0
+    if all_keys_eq(row, row_prev):
+        assert row.value_prev == 0
+
```
Comment (ed255) [inline src/zkevm_specs/state.py:339]: I think the TODO in the next line can now be removed :)

File: src/zkevm_specs/state.py [hunk 2]
```diff
@@ -341,8 +349,11 @@ def check_tx_access_list_account(row: Row, row_prev: Row):
     assert row.field_tag() == 0
     assert row.storage_key() == 0
 
+    # 9.1 When keys change, value_prev(is_warm_prev) must be False
+    if all_keys_eq(row, row_prev):
+        assert row.value_prev == 0
+
```
Comment (ed255) [inline src/zkevm_specs/state.py:355]: I think the TODO in the next line can now be removed :)

File: src/zkevm_specs/state.py [hunk 3]
```diff
@@ -518,6 +532,19 @@ def keys_rwc_to_limbs_in_order(row: Row) -> List[FQ]:
...
+        reversible_rws = [
+            Tag.Storage,
+            Tag.Account,
+            Tag.TxRefund,
+            Tag.TxAccessListAccount,
+            Tag.TxAccessListAccountStorage,
+            Tag.TxRefund,
+        ]
+        if row.tag() in reversible_rws:
+            assert row.value_prev == row_prev.value
```
Comment (ed255) [inline src/zkevm_specs/state.py:546]: Isn't this constraint already enforced in `check_account` and `check_storage` for `Tag.Account` and `Tag.Storage`?

On a second note, writing this constraint only for specific tags is more complex than writing for all tags. Maybe we can have this constraint for all tags (even when it's not needed), and just make sure the witness we generate is valid.

**Comment by lispc** [comment]
thanks for the review!

I am thinking about using RLC these several days. Maybe it can brings some optimization / difference.. Updates soon.

File: src/zkevm_specs/state.py (hunk 3 above)
Comment (CPerezz) [reply src/zkevm_specs/state.py:546]: I agree completely with that!

If we can make a generic constraint that works over all of them at once that would make it much simpler.

**Review by CPerezz — COMMENTED** [review]
As for the RLC of the Tag column, avoinding to have a column for `value_prev` I would explore the idea but start with the easiest implementation problably.

We have time to optimize, and also, considering that we will refactor the RLC to work with splitted words (hi,lo) adding more RLC into our designs now would probably add tech-debt for the close future.

**Comment by ed255** [comment]
> As for the RLC of the Tag column, avoinding to have a column for `value_prev` I would explore the idea but start with the easiest implementation problably.

I agree that it's OK to start with an easier solution first (which may not be optimal) and optimize in a future iteration.  Nevertheless I would defer this decision to the person specifying / implementing the circuit :)

> We have time to optimize, and also, considering that we will refactor the RLC to work with splitted words (hi,lo) adding more RLC into our designs now would probably add tech-debt for the close future.

AFAIU, the RLC for tables is not the same as the RLC for words.  So we may remove the word-RLC but keep other RLCs.  In particular, the RLC for the rw_table only requires a randomness shared between the EVM circuit and state circuit, and that's easier to achieve than the randomness required for words (which involves many circuits).

**Comment by ed255** [comment]
> But there is a missing part in this design: the committed_value in Rw::Account. Without this, the first value_prev of Rw::Account cannot be constraint or calculated. My current constaint of value_prev [is like this](https://github.com/scroll-tech/zkevm-circuits/pull/155/files#diff-3c34e18842d60643d37a4c79494077f57237418db3eb5b2539827baa23859d60R138).
> 
> So my conclusion: adding the real value_prev column may not be needed. But in order to constraint prev_value used in evm circuit lookup, we have to add Rw::Account::committed_value in bus-mapping / evm circuit first. So i guess this PR can be hold for a while before [privacy-scaling-explorations/zkevm-circuits#557](https://github.com/privacy-scaling-explorations/zkevm-circuits/issues/557) is fixed.

The current spec contains the committed_value for Account ops in the state circuit, which is used in the check_account constraints:
Table spec: https://github.com/privacy-scaling-explorations/zkevm-specs/blob/b436982c60aa38e91028212d0abe21a50e405a2a/specs/tables.md?plain=1#L70-L72
state circuit account op constraint: https://github.com/privacy-scaling-explorations/zkevm-specs/blob/b436982c60aa38e91028212d0abe21a50e405a2a/src/zkevm_specs/state.py#L313

I believe what's missing is the corresponding update in the EVM circuit spec, and then updating the implementation (bus-mapping, evm circuit, state circuit) to add the account committed_value which is currently missing.

**Comment by CPerezz** [comment]
>AFAIU, the RLC for tables is not the same as the RLC for words. 

I'm not sure TBH. Would need to look it up. It makes sense what you say since we encode columns in the tables. Therefore, we will have 2x columns or rows due to the split in hi/lo limbs for the Words but the mechanism to encode everything will still be untouched.

Thanks @ed255 :smile:

**Pull request closed without merging**